source/request-response/subscription_manager.c

```diff
@@ -0,0 +1,627 @@
+    struct aws_rr_subscription_record *record = aws_mem_calloc(allocator, 1, sizeof(struct aws_rr_subscription_record));
+    record->allocator = allocator;
+
+    aws_byte_buf_init_copy_from_cursor(&record->topic_filter, allocator, options->topic_filter);
+    record->topic_filter_cursor = aws_byte_cursor_from_buf(&record->topic_filter);
+
+    aws_hash_table_init(
+        &record->listeners,
+        allocator,
+        4,
+        s_aws_hash_subscription_listener,
+        s_aws_subscription_listener_hash_equality,
+        NULL,
+        s_aws_subscription_listener_destroy);
+
+    record->status = ARRSST_NOT_SUBSCRIBED;
+    record->pending_action = ARRSPAT_NOTHING;
+
+    record->type = options->type;
+
+    return record;
+}
+
+static void s_subscription_record_unsubscribe(
+    struct aws_rr_subscription_manager *manager,
+    struct aws_rr_subscription_record *record,
+    bool shutdown) {
+
+    bool currently_subscribed = record->status == ARRSST_SUBSCRIBED;
+    bool currently_subscribing = record->pending_action == ARRSPAT_SUBSCRIBING;
+    bool currently_unsubscribing = record->pending_action == ARRSPAT_UNSUBSCRIBING;
+
+    /*
+     * The difference between a shutdown unsubscribe and a normal unsubscribe is that on a shutdown we will "chase"
+     * a pending subscribe with an unsubscribe (breaking the invariant of never having multiple MQTT operations
+     * pending on a subscription).
+     */
+    bool should_unsubscribe = currently_subscribed && !currently_unsubscribing;
+    if (shutdown) {
+        should_unsubscribe = should_unsubscribe || currently_subscribing;
+    }
+
+    if (!should_unsubscribe) {
+        AWS_LOGF_DEBUG(
+            AWS_LS_MQTT_REQUEST_RESPONSE,
+            "request-response subscription manager - subscription ('" PRInSTR
+            "') has no listeners but is not in a state that allows unsubscribe yet",
+            AWS_BYTE_CURSOR_PRI(record->topic_filter_cursor));
+        return;
+    }
+
+    struct aws_protocol_adapter_unsubscribe_options unsubscribe_options = {
+        .topic_filter = record->topic_filter_cursor,
+        .ack_timeout_seconds = manager->config.operation_timeout_seconds,
+    };
+
+    if (aws_mqtt_protocol_adapter_unsubscribe(manager->protocol_adapter, &unsubscribe_options)) {
+        int error_code = aws_last_error();
+        AWS_LOGF_ERROR(
+            AWS_LS_MQTT_REQUEST_RESPONSE,
+            "request-response subscription manager - sync unsubscribe failure for ('" PRInSTR "'), ec %d(%s)",
+            AWS_BYTE_CURSOR_PRI(record->topic_filter_cursor),
+            error_code,
+            aws_error_debug_str(error_code));
+        return;
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_MQTT_REQUEST_RESPONSE,
+        "request-response subscription manager - unsubscribe submitted for ('" PRInSTR "')",
+        AWS_BYTE_CURSOR_PRI(record->topic_filter_cursor));
+
+    record->pending_action = ARRSPAT_UNSUBSCRIBING;
+}
+
+/* Only called when shutting down the client */
```

Comment:
Trivial: client->manager

Reply:
Actually that's intentional.  The (subscription manager) function is only called when the whole request-response client is getting shut down.